### PR TITLE
Allow filtering by unseen recordings

### PR DESCRIPTION
**Pull request by macobo**
- Add model for session recording viewed
- Save view when querying for session recording data
- Send information to FE about whether session has been viewed
- Allow filtering by "recording unseen"
- Rename property

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

### CODE REVIEW

File: posthog/models/session_recording_event.py [hunk 1]
```diff
@@ -19,3 +19,14 @@ class Meta:
     distinct_id: models.CharField = models.CharField(max_length=200)
     session_id: models.CharField = models.CharField(max_length=200)
     snapshot_data: JSONField = JSONField(default=dict)
+
+
+class SessionRecordingViewed(models.Model):
```
Comment (paolodamico) [inline posthog/models/session_recording_event.py:24]: I would inherit this from `UUIDModel` to use Michael's UUIDT model and avoid tech debt with the primary keys thing.

File: posthog/queries/sessions/session_recording.py [hunk 2]
```diff
@@ -94,20 +103,32 @@ def filter_sessions_by_recordings(
...
-        session["session_recording_ids"] = [
-            recording["session_id"] for recording in session_recordings if matches(session, recording)
-        ]
+        session["session_recordings"] = list(
+            collect_matching_recordings(session, session_recordings, filter, viewed_session_recordings)
+        )
 
     if filter.limit_by_recordings:
-        sessions_results = [session for session in sessions_results if len(session["session_recording_ids"]) > 0]
+        sessions_results = [session for session in sessions_results if len(session["session_recordings"]) > 0]
```
Comment (paolodamico) [inline posthog/queries/sessions/session_recording.py:116]: I think we have two separate for loops here iterating over the same content, maybe worth optimizing?

File: posthog/queries/sessions/session_recording.py [hunk 3]
```diff
@@ -94,20 +103,32 @@ def filter_sessions_by_recordings(
     max_ts = max(it["end_time"] for it in sessions_results)
 
     session_recordings = query(team, min_ts, max_ts, filter)
+    viewed_session_recordings = set(
```
Comment (paolodamico) [inline posthog/queries/sessions/session_recording.py:106]: I might be missing something but wouldn't it be better to delegate the heavy lifting of matching viewed session recordings to Postgres? As in get the list of sessions first and then do the select on the DB as opposed to getting the list of all viewed recordings every time?

File: posthog/api/event.py [hunk 4]
```diff
@@ -299,11 +300,17 @@ def session_events(self, request: request.Request, *args: Any, **kwargs: Any) ->
     # /event/session_recording
     # params:
     # - session_recording_id: (string) id of the session recording
+    # - save_view: (boolean) save view of the recording
```
Comment (paolodamico) [inline posthog/api/event.py:303]: Nit: the name confused me at first. I think something like `record_watched` or something like that would make it more clear that you're storing that the user has already watched a given session recording.

**Review by paolodamico — APPROVED** [review]
A few minor comments but nothing blocking, lgtm!

File: posthog/queries/sessions/session_recording.py (hunk 3 above)
Comment (macobo) [reply posthog/queries/sessions/session_recording.py:106]: This wouldn't work for clickhouse as the two tables are stored in different dbs.

File: posthog/queries/sessions/session_recording.py (hunk 2 above)
Comment (macobo) [reply posthog/queries/sessions/session_recording.py:116]: This wouldn't really speed things up as and would be slightly unclearer.

File: posthog/models/session_recording_event.py (hunk 1 above)
Comment (macobo) [reply posthog/models/session_recording_event.py:24]: Not sure what the benefit is in this case - we're doing lookups via an index anyways.

File: posthog/api/event.py (hunk 4 above)
Comment (macobo) [reply posthog/api/event.py:303]: Problem with `record_viewed` is that it sounds like another filter - whereas this mutates state.

**Pull request merged**